Bug798673 - Balance Limits

Bob-IT opened the pull request:
This is a basic implementation of I think what Chris was looking at in this bug. Still needs a bit of work but posting here for some feedback and questions. I have added a KVP entry for a balance limit and entry for whether to use 'greater than'. 

1) Is there an account that would require a minimum and maximum limit at the same time?
2) I have added another tab to the account dialog, what to call it keeping in mind other properties may be added later?
3) Use of positive numbers makes sense in that 100 is greater than 90 but with negative numbers it may be confusing that -90 is greater than -100 so I have transposed the numbers for gnc_numeric_compare. This is probably wrong.
4) Dialog text needs to be better with appropriate tool tips, suggestions?
5) Not sure about the icon?

It would be possible to add a different icon when close to the limit, say 10%, but then some one would ask if it could be changed so maybe if wanted a warning balance limit could be added. This could possibly display an 'orange' icon and the limit a 'red' one.

Tests will need to to added which I will do after feedback on changes required.

@christopherlam
Have a spin and see what you think, I would imagine that there could be a report that would list all the accounts with the balance set and maybe indicate how close the balance is to it.


![balance-limit](https://user-images.githubusercontent.com/15702727/205666132-b532e4f8-e1a9-4bbb-bf48-3ba0975638cb.png)

jralls commented:
> Is there an account that would require a minimum and maximum limit at the same time?

Sure, a demand account that a user wants to be warned about if it's getting close to its minimum balance below which the bank will charge a fee or has enough that the user would want to transfer some to an account with a higher return.

> I have added another tab to the account dialog, what to call it keeping in mind other properties may be added later?

Depends on what you're contemplating adding later. For the current usage I'd call it Limits or Warning Limits. If you have in mind properties unrelated to that maybe More Properties or Other Properties.

> Use of positive numbers makes sense in that 100 is greater than 90 but with negative numbers it may be confusing that -90 is greater than -100 so I have transposed the numbers for gnc_numeric_compare. This is probably wrong.

The computation looks correct to me but I'm concerned that some users might have trouble with the `Greater than`/`Less than` radio buttons. You're thinking of them in terms of a number line where -100 < -90, but it's equally valid to think in terms of magnitude and `gnc_numeric_abs(-100) > gnc_numeric_abs(-90)`: A user whose checking account provides overdraft protection might set a negative balance limit to warn them when they're getting close to the maximum overdraft. They might think that they want to check `Greater than` because they don't want to go "over" the overdraft limit, or they might think in the number-line sense and check `Less than`. 

For display purposes you also need to check GNC_PREF_REVERSE_ACCOUNT_CREDIT and GNC_PREF_REVERSE_ACCOUNT_INCOME_EXPENSE so that the limit and direction the user enters reflects the sign of the numbers they see in the register. With the number-line meaning of greater and less that will also require the radio buttons and their interpretation to flip if a REVERSE_ACCOUNT preference applies to the account.

I think it would be simpler to make the radio buttons indicate the magnitude instead of number-line meaning of greater and less. Lots of our users don't think mathematically so we should avoid using mathy terms. What do you think of making the labels "More than" and "Less than" with tool tips and translator comments explaining that it's more or less in the magnitude sense, i.e. that -100 is more than -90 just like 100 is more than 90?

christopherlam commented:
Nice start. I'd thought the balance limit would be implemented as a `(minimum <= balance <= maximum)` test. The [minimum, maximum] intervals would be input via GNCAmountEdit, and the signs would be natural signs. i.e. the suspense accounts would be [0,0]; the credit card would be [-2000,0]. An account which accumulates only rounding errors (e.g. I handle GSTs on sales/purchases but the taxman only wants whole digits) may have limits [-10,10].

Alternatively we could have a single positive maximum limit, and the test is `abs(balance) <= limit`.

Bob-IT commented:
Thank you for the feedback, as there could be a higher and lower limit I changed the dialog to have two entries as seen below.

![balance-limit2](https://user-images.githubusercontent.com/15702727/206863406-81ec406c-fb6a-4610-bead-8e5e69e3f6a7.png)

I was originally going to hide the option if an account had children but decided to add another setting to include balances of child accounts in the test. The inclusion of the preference settings was tricky but I think it works OK and I have also changed the icons displayed.
![balance-limit](https://user-images.githubusercontent.com/15702727/206863399-d438afb1-6f48-4b97-8da8-8df6283ea7e2.png)

jralls commented on libgnucash/app-utils/gnc-ui-balances.c:
```diff
@@ -250,6 +250,75 @@ gnc_ui_account_get_reconciled_balance_as_of_date (Account *account,
+        balance_limit = gnc_numeric_neg (balance_limit);
+    }
+
+    /** Returns 1 if a>b, -1 if b>a, 0 if a == b  */
```
Shouldn't be a Doxygen comment.

Bob-IT replied:
Copy and pasted, changed.

jralls commented:
That's much clearer. I like the icons in the accounts page.

Can you bring yourself to add a section describing the new tab at https://github.com/Gnucash/gnucash-docs/blob/master/C/manual/ch_Account-Actions.xml#L729?

Bob-IT commented on gnucash/gtkbuilder/dialog-account.glade:
```diff
@@ -1682,7 +1682,155 @@
+
+i.e.
+A balance of -90 will show icon if limit set to -100
+A balance of 100 will show icon if limit set to 90
```
Changed

Bob-IT commented:
> Can you bring yourself to add a section describing the new tab at https://github.com/Gnucash/gnucash-docs/blob/master/C/manual/ch_Account-Actions.xml#L729?

I will have a look, not sure I am setup to pull the docs.

christopherlam commented:
I haven't tested this thoroughly, but I'm _very_ nervous when (1) including subaccounts (2) handling `gnc_reverse_account` balances... 

Are all balance signs tested properly? 
Are they stored unreversed, and reversed for GNCAmountEdit display/input only? 
What if Sign-Reversal global pref is changed while the dialog is active? 
What if an Asset account has a Liability subaccount?

We won't want to repeat the budget amount's signs' bug. Hence in my comment I'd mentioned using _natural_ signs (with the usual UX issues), without considering inclusion of subaccounts (which is nice but rather difficult to test 100%).

Bob-IT commented:
> Are all balance signs tested properly?
Do not know what you mean?


> Are they stored unreversed, and reversed for GNCAmountEdit display/input only?
They should be, the values saved should be when the preference is none, they are reversed/unreversed for display.


> What if Sign-Reversal global pref is changed while the dialog is active?
Not sure you can but will try, if so I think I just need to record the `gnc_reversed` value on load and use that for save.


> What if an Asset account has a Liability subaccount?
Should not matter, the limits are for that account against the existing balance for that account so the balance should follow the account setting.

I do not think there should be an issue with sub accounts as stated above but will try and do some more testing.

christopherlam commented:
Also need to consider the case whereby subaccount has a different currency.
It means the parent account balance test will depend on currency
fluctuations.

On Mon, 12 Dec 2022, 5:06 am Bob-IT, ***@***.***> wrote:

> Are all balance signs tested properly?
> Do not know what you mean?
>
> Are they stored unreversed, and reversed for GNCAmountEdit display/input
> only?
> They should be, the values saved should be when the preference is none,
> they are reversed/unreversed for display.
>
> What if Sign-Reversal global pref is changed while the dialog is active?
> Not sure you can but will try, if so I think I just need to record the
> gnc_reversed value on load and use that for save.
>
> What if an Asset account has a Liability subaccount?
> Should not matter, the limits are for that account against the existing
> balance for that account so the balance should follow the account setting.
>
> I do not think there should be an issue with sub accounts as stated above
> but will try and do some more testing.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Gnucash/gnucash/pull/1480#issuecomment-1345656494>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPCMPWZO3GRL7DJQS3REWDWMY637ANCNFSM6AAAAAASUNCMGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Bob-IT commented:
> Also need to consider the case whereby subaccount has a different currency. It means the parent account balance test will depend on currency fluctuations.

I do not think so, as I said before the limits are against the account. Should not matter what the sub accounts are as long as the balance function is correct and gives a valid balance for the account with limits set. I would expect any body with accounts setup as you suggest would realise the balance of the parent account will change with currency fluctuations.

Bob-IT commented on gnucash/gnome-utils/dialog-account.c:
```diff
@@ -892,6 +981,27 @@ gnc_common_ok (AccountWindow *aw)
+        gint compare = gnc_numeric_compare (higher_balance_limit,
+                                            lower_balance_limit);
+
+        if (compare != 1)
+        {
+            const char *message = _("The lower balance limit must be less than the higher limit.");
+            gnc_error_dialog (GTK_WINDOW(aw->dialog), "%s", message);
+            LEAVE("invalid balance limit");
+            return FALSE;
+        }
```
@christopherlam  How about...

        if ((compare == 0) && (!gnc_numeric_zero_p (higher_balance_limit)))
        {
            const char *message = _("Balance limits must be different unless they are both zero.");
            gnc_error_dialog (GTK_WINDOW(aw->dialog), "%s", message);
            LEAVE("invalid balance limit, both the same but not zero");
            return FALSE;
        }
        else if (compare == -1)
        {
            const char *message = _("The lower balance limit must be less than the higher limit.");
            gnc_error_dialog (GTK_WINDOW(aw->dialog), "%s", message);
            LEAVE("invalid balance limit, lower limit not less than upper");
            return FALSE;
        }

christopherlam replied:
This looks like overengineering... I'll be quite happy to not worry about it and set limits as 0 <= balance <= 0.01...

Bob-IT replied:
> This looks like overengineering...

I do not think so, any way I have added it so you can specify 0,0 for the limits. As I understand it, you are really interested in highlighting that the balance is not zero so I have added changing the icon to 'dialog-warning' as can be seen in image...

![balance-limit3](https://user-images.githubusercontent.com/15702727/208240437-8e2b7d3e-4656-41eb-b94c-439225963894.png)

jralls commented:
Looks great, thanks!
